File: 2022/src/main/kotlin/days/Day04.kt
package days

class Day04 : Day(4) {

    private val input = inputList
        .map { line ->
            line.split(",")
                .map { it.split("-") }
                .map { (a, b) -> a.toInt()..b.toInt() }
        }.map { (a, b) ->
            a to b
        }

    override fun part1() : Any {
        return input.count {(a, b) ->
            (a.first in b && a.last in b) || (b.first in a && b.last in a)
        }
    }

    override fun part2() : Any {
        return input.count {(a, b) ->
            (a.first in b || a.last in b) || (b.first in a || b.last in a)
        }
    }
}

File: 2017/src/days/day04.rs
use itertools::Itertools;
use std::collections::HashSet;

use super::day::Day;
use crate::util::solution::Solution;

pub struct Day04;

impl Day for Day04 {
    fn part1(input: Vec<String>) -> Solution {
        let result = input
            .iter()
            .filter(|&line| {
                let words = line.split_whitespace();
                let unique_words: HashSet<&str> = HashSet::from_iter(words.clone());
                words.count() == unique_words.len()
            })
            .count();

        return Solution::I32(result.try_into().unwrap());
    }

    fn part2(input: Vec<String>) -> Solution {
        let result = input
            .iter()
            .filter(|&line| {
                let words = line.split_whitespace();
                let unique_words: HashSet<&str> = HashSet::from_iter(words.clone());
                let sorted_words: HashSet<String> =
                    HashSet::from_iter(words.clone().map(|w| w.chars().sorted().join("")));
                words.clone().count() == unique_words.len() && words.count() == sorted_words.len()
            })
            .count();

        return Solution::I32(result.try_into().unwrap());
    }
}
